### DISCUSSION TITLE
using non-ASCII characters in input sequence (m17n?)

### DISCUSSION THREAD

**Discussion by silmeth** [fcitx/fcitx5#999] [Q&A] [answered]
Hi, I’ve recently tried to move from ibus to fcitx5 and thus I tried to migrate my [Cyrillic ibus-table](https://github.com/silmeth/ms-cyrillic-ibus-table) input method to m17n so that I could use it in fcitx. The first attempt can be seen [in my dotfiles on gitlab](https://gitlab.com/silmeth/configs/-/blob/master/keyboard/.m17n.d/interslavic-cyrillic.mim).

But I have a problem. I use Polish keyboard as my main default keyboard layout – and Polish layout lets me input characters like ⟨ż⟩ via <kbd>AltGr</kbd> + <kbd>z</kbd>, or ⟨ę⟩ via <kbd>AltGr</kbd> + <kbd>e</kbd> – and I did use those in the input sequences to transform them into Cyrillic ⟨ж⟩ and ⟨ѧ⟩ respectively.

So I added rules like these to my mapping:

```sexp
    ("Ż" ?Ж)
    ("ż" ?ж)
```

but they don’t work – when I switch the input method to my Cyrillic, hitting <kbd>AltGr</kbd> + <kbd>z</kbd> produced unchanged Polish ⟨ż⟩.

The [m17n data format spec](https://www.nongnu.org/m17n/manual-en/m17nDBFormat.html#mdbIM) says about `KEYSEQ`:

> `MTEXT` in the definition of `KEYSEQ` consists of characters that can be generated by a keyboard. Therefore `MTEXT` usually contains only `ASCII` characters. However, if the input method is intended to be used, for instance, __with a West European keyboard, `MTEXT` may contain Latin-1 characters__.

so I’d expect this to work. Is this a bug in fcitx5, a consciously enforced limitation? Is there a way around it?

I also tried making rules with modifier keys explicitly myself, eg:
```sexp
    ((A-z) ?ж)
```
and this lets me use the left <kbd>Alt</kbd> but not the right one, I also tried:
```sexp
    ((G-z) ?ж)
```
(`G-` is the m17n prefix for AltGr) but this does nothing with my Polish keyboard layout.

**Comment by wengxt** [comment, marked as answer]
While it's totally up to you that whether to use m17n to implement it, have you checked the table engine in fcitx5? (It's a part of fcitx5-chinese-addons). 

Use libime_tabledict to convert it 
[interslavic-cyrillic.txt](https://github.com/fcitx/fcitx5/files/14813882/interslavic-cyrillic.txt)
```
libime_tabledict interslavic-cyrillic.txt interslavic-cyrillic.main.dict
```

And save it to ```~/.local/share/fcitx5/table/```

Save this file to ```~/.local/share/fcitx5/inputmethod/``` you may want to update the informational field to some other value like lang code & name.
[interslavic-cyrillic.conf](https://github.com/fcitx/fcitx5/files/14813884/interslavic-cyrillic.txt)

**Reply by wengxt** [reply]
You can also check some other examples in fcitx5-table-other on how it works

e.g.
https://github.com/fcitx/fcitx5-table-other/tree/master/tables/pa

**Reply by silmeth** [reply]
Thanks! This does work. :)

I couldn’t find any documentation how to make a new table layout, only found [this outdated(?) wiki article](https://fcitx-im.org/wiki/How_to_make_your_own_table-based_input_method) which said that input keys must be ASCII – so I assumed that still likely holds. I see it’s not the case anymore.

So this solves my particular problem.

Still the question about m17n holds – shouldn’t non-ASCII input characters in the key seq work there?

**Reply by wengxt** [reply]
m17n was moved to https://github.com/fcitx/fcitx5-m17n/issues/4

**Reply by silmeth** [reply]
Thanks!

**Reply by wengxt** [reply]
@silmeth I made some investigation and commented on https://github.com/fcitx/fcitx5-m17n/issues/4, please take a look.
